tactics/hints.ml

```diff
@@ -1000,7 +1000,7 @@ let make_extern pri pat tacast =
   (hdconstr,
    { pri = pri;
      pat = Option.map (fun p -> SyntacticPattern p) pat;
-     name = None;
+     name = name;
```

Comment:
This is wrong: this name is a kernel-facing name, but you're using it for a non-kernel object. You should use another nametab for that.

Reply:
if it's a separate nametab won't it make Remove Hint ambiguous? ie in `Remove Hint x`, `x` could be a globref or it could be a extern

Reply:
It certainly compiles.  The name specified in `Hint Extern` has to be an existing qualid or you get an error message.  That's not ideal, but it's a step forward to be able to give a name at all.  `Hint Resolve` sets the name automatically from the theorem name.  It seems reasonable to expect that most Extern hints would have fairly obvious existing qualids to use for the name.  If necessary, they could create a qualid with `Definition` beforehand.  If I create a new namespace, then this record would have to have a different field to store the new type of name--either wrapping the current `name` field with a constructor or creating an additional field.

I don't want to try to enforce uniqueness on the name at this time--that would not be backwards compatible in some cases.  You'd then need syntax to specify the name for each of the hint commands.  The effort is higher and the value may be lower.  As I discuss in #19816.

Reply:
> if it's a separate nametab won't it make Remove Hint ambiguous?

The command is `Remove Hints`, with an "s".  Current behavior is that it removes all hints with the given name.  That's not what I would have done.

Reply:
> It seems reasonable to expect that most Extern hints would have fairly obvious existing qualids to use for the name.

doubt

Reply:
> It seems reasonable to expect that most Extern hints would have fairly obvious existing qualids to use for the name.

I agree, and I don't think it's such a bad idea to use references here, as long as the idea remains to identify hints using _existing_ references (and not to create ad-hoc ones that don't have a corresponding kernel object).

We've recently started to uniquely identify our `Extern` hints using some marker tactic. We found that relying on a references was fairly natural, and (at least for us) there was indeed a natural choice of an existing definition name at least 99% of the time. In other situations, we simply created a dummy definition next to the hint declaration, which wasn't a big problem.

Reply:
This is ambiguous, though. The name could refer to both an Extern hint and some Resolve / Immediate hint generated by the reference.

Reply:
That's true, but isn't that already the case with other forms of hints?

Reply:
It's definitely already ambiguous. In Coq's defense, the command `Remove Hints` is itself phrased in an ambiguous way and could well be intended to mean "remove all hints associated with the reference(s)".

```coq
Create HintDb test discriminated.
Class Bla := BLA {}.
Definition bla : Bla := BLA.
Hint Unfold  bla : test.
Hint Resolve bla : test.
Hint Immediate bla : test.
Print HintDb test. (* 3 hints *)
Remove Hints bla : test
Print HintDb test. (* no hints *)
Print HintDb test. (* no idea why I need two [Print HintDb] but the last one never shows up in 8.19 *)
```

Reply:
> no idea why I need two [Print HintDb]

You left out the period on the `Remove Hints`.

Reply:
We could move toward prohibiting duplicate names _within each hintDb_, perhaps like this:
- Limit names for `Extern Hint`s to a simple `ident` instead of a `qualid`.  If the `ident` matches another `Hint Extern` name or the last component of a hint with a `qualid` name, give an error.  So there would be two types of names for hints.
- For non-Extern hints, give a warning for a duplicate and later an error.  Users would have to give a name explicitly as a simple `ident` if the qualid is already used.  (Giving the user a choice to present a simple `ident` _or_ a `qualid` seems like overkill.  Syntax for that would be baroque.)
- Adjust `Remove Hints` and `Hint Cut` code to work the same with `qualid` or `ident` names.
- Users will need to adjust their `Remove Hints` and `Hint Cut`s where they rely on shared use of the `qualid`.  These commands could generate warnings for non-unique references to help users make these changes.  I expect users don't rely heavily on non-unique references.

IMO
- It would be strange for Hint commands to define qualified names such as `Nat.something`, which should be defined in the `Nat` module.
- I would not require names for Extern hints.  I expect most users don't use `Remove Hints` or `Hint Cut`, so why force them to change?
- Using the qualid for non-Extern hints is convenient.  I don't want to force users to explicitly name everything--IMO most users don't need this.
- Being able to reuse hint names across hintDbs seems appropriate.

WDYT?

Reply:
Non-qualified names are terrible from a modularity point of view, e.g. just look at the mess with absolute names in hint databases. I think that the problem comes from the Remove command that was implemented quickly as a way to work around badly designed hint declarations in foreign libraries. If anything, we should just extend it to support the dual of hints declarations, i.e. we should be able to remove hints depending on their Resolve / Immediate / Unfold status.

Reply:
> - I would not require names for Extern hints. I expect most users don't use Remove Hints or Hint Cut, so why force them to change?

Even if there is no use for removing Extern hints withing a library, it does not mean that client code may not require to do so (e.g., for performance reason). So, in my opinion, we should ensure that it is always possible to remove Extern hints, even if that means forcing users to give them names. Of course, this would need to follow a standard deprecation cycle.

Reply:
@rlepigre Maybe you didn't fully get my point: Let users decide if they want to provide hint names and never force them to provide them.  What you say above is an argument that libraries should always provide hint names for Hint Extern so that they can be removed.  OTOH, if it's optional, it will be easier to inadvertently forget to give names in libraries--though readily fixed when discovered.

Are your hints scattered across many files (like `core`) or do you create your own hintDb and add all its hints in a single .v file?  The latter would allow complete control over the ordering of hints, which may affect the run times of proof searches.

> @ppedrot Non-qualified names are terrible from a modularity point of view

OK, we can stick with providing an (existing??) qualid for `Hint Extern` names.  The command will verify that the name is not already in use within the hintDb.

> we should be able to remove hints depending on their Resolve / Immediate / Unfold status.

That could be done.  Note there are also Resolve -> and Resolve <- (which haven't worked when I've tried to create examples, but I do see uses of the construct in the standard library, so something I'll get back to in a while).  However, the hint names are also used in `Hint Cut` expressions.  I think including Resolve, etc. in cut expressions would be hideous.

One alternative is to mangle the qualid, so `Resolve foo` creates or reuses the qualid `foo_Resolve`.  But that could force immediate changes to `Remove Hints` and `Hint Cut`.  Not as good as I'd like, but the best I can think of at the moment.

Reply:
@jfehrle no, I think I got it right, I am really arguing for forcing users to always tag Extern hints with a reference (or some other kind of identiifer). By the way, I still think that using a reference is the best option at this point, even if that means that `Remove Hints` continues to remove all hints associated to a particular reference at once. This behaviour is not completely unreasonable, and would suffice to give users the control they are currently lacking.

Reply:
I'm also in favour of making the reference mandatory, otherwise this completely misses the point of the Remove Hints command. If you're in your own code it's easy to fix bad hint declaration, much less so when using an external library.

Regardless of this point, given the above discussion I'm even more convinced that we want a specific namespace for extern hints. Remove hints was not very well designed, but that's no excuse to stack more hacks atop of the implementation. We should really have a symmetric version of hint arguments to be able to selectively remove hints in a database, rather than relying on fragile and ambiguous constant names.

Reply:
OK, you both want the name mandatory for `Hint Extern`.  Maybe the way to go is to require an explicit name for all hints--@ppedrot that seems to be your drift.  Is that correct?  The initial release can give a warning if no name is provided to allow a reasonable migration.  If the new name is not set, provide the existing behavior for `Remove Hints` and `Hint Cut`.

We'd have to add names in all the libraries.  If any of them don't match the current names, that could require users to make immediate repairs to their scripts for `Remove Hints` and `Hint Cut`.

> We should really have a symmetric version of hint arguments

I don't want to see `Remove Hint Extern foo` (if that's what you mean) because `Hint Cut Extern foo Resolve bar Resolve <- bas` is uglier than sin, verbose and not easy to read.  Using the same explicit name is simple and clear.

BTW, we should create a new `Remove Hint` command.

Reply:
> Is that correct?

No, the only kind of hints that needs a name is Extern.

> Hint Cut

I don't know why you started mentioning Hint Cut in the above discussion. Hint Cut is not a hint entry, that's a database modifier. Currently AFAIK there is no way to remove a cut, and in any case negating it would be performed by a command with a similar syntax as Hint Cut. These things are regexps, it should be possible to negate them, but the current implementation will need a serious rework before doing that efficiently.

> I don't want to see Remove Hint Extern foo

The could be some syntactic sugar where `Remove Hint foo` means anything that matches the name foo in the db with a warning in case of ambiguity, with a fallback to the qualified Extern / Immediate / Resolve command if needed.

> Hint Cut Extern foo Resolve bar Resolve <- bas

As already mentioned, there is not Hint Cut in there, but furthermore the `<-` syntax is syntactic sugar. IIRC nowadays it generates two intermediate lemmas from the input, so you can (and should) just remove those ones. The syntax should really match what we have in the db, i.e. something along the lines of the hint_ast ADT.

If you're really concerned about legibility, there could be a lightweight syntax with a list of optional desambiguation annotations, e.g.
```
Remove Hints foo bar [immediate] baz [extern] qux [unfold, resolve].
```

Keep in mind that Resolve Hints is essentially a hack to work around ill-designed libraries, so it's not something you use very often.

Reply:
> Hint Cut is not a hint entry, that's a database modifier.

Right.  I'm not suggesting that there should be a way to remove Hint Cuts.  Rather, Hint Cut relies on the names provided in the hints.  If we're going to provide a way to uniquely identify hints in Hint Remove, we should do the same for Hint Cut.  Requiring a unique name for every hint is clean and simple, better than applying more band-aids.  Having two types of names for hints is awkward.

Note that fixing Remove Hints for non-Extern hints is already more than I had planned to do in this PR.  I did intend to do something that works for Hint Cut.

> IIRC nowadays [the <- syntax] generates two intermediate lemmas from the input, so you can (and should) just remove those ones.

So far I've not gotten any examples to work so I can't verify that, e.g.

```
Hint Resolve <- plus_n_O : db.   (* forall n : nat, n = n + 0 *)
(* Error: Cannot recognize a statement based on iff. *)
```

Furthermore, this hint is useful right to left but problematic from left to right.  Generating both lemmas seems a bad idea.  And users might want to include applying the theorem in addition to rewriting with it in their hintDb.

Reply:
These work.  They generate separate hints:

```
Hint Resolve -> and_comm.
Hint Resolve <- and_comm.
```

Reply:
These hints are bound to two autogenerated kernel definitions `and_comm_proj_l2r` and `and_comm_proj_r2l`, so this is indeed what I was mentioning before: you should remove those instead of introducing a dedicated syntax.

Reply:
You mean `Remove Hints and_comm` should remove both?  Current behavior is that this is a no-op and `Remove Hints and_comm_proj_l2r` removes the hint.  This could be documented rather than changing the behavior.

`Hint Constructors nat` is similar, removing `nat` is a no-op and removing `S` works.  Since it can create any number of hints, introducing explicit names for each constructor doesn't make sense.  Maybe this can be documented instead of changing the behavior.

It may helpful for `Remove Hints` to give a warning if doesn't remove anything.

> introducing a dedicated syntax

Which dedicated syntax?  My suggestion of adding explicit names everywhere or your `Remove Hints foo bar [immediate] baz [extern] qux [unfold, resolve].`?

I tried to construct a `Hint Cut` example that actually does something without success.  I always see empty paths in `Hints.path_derivate` (the second one).  I expect it uses the same names as `Remove Hints`.  `Hint Cut` accepts `and_comm` and `nat`.  Maybe the command could check the type of object more closely and report an error at least for `nat`.

Still thinking about what should be in this PR.

Reply:
I believe `Hint Extern name => tac` should introduce an Ltac (1 or 2) definition `name := tac` and that `Remove Hints` should take a `glob_ref | tactic_name`. Isn't that doable?

Reply:
How would the command know whether to use Ltac 1 or 2? Different namespaces and different parsing.  I don't see the benefit of creating Ltac tactics for elementary tactics such as "split".  It would be a  bit more overhead wrapping split as an Ltac tactic. 
 Do you think Ltac tactics are really the predominant use case?

Reply:
I would not make the name required, either.  I expect many users don't use Remove Hunt at all.

Reply:
And we'd need to move the hint cost elsewhere in the syntax.

BTW, I will get back to this PR, but it may not be until after the holidays and other distractions.